### PR TITLE
shrinking binaries post

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

File: content/2023-06-02-shrinking-binaries.rst [hunk 1]
```diff
@@ -0,0 +1,358 @@
+Shrinking a Shared Library
+##########################
+
+:date: 2023-06-02
+:category: mozilla
+:lang: en
+:authors: Serge Guelton
+:summary: What is the effect of different compiler flags on the size of a shared
+          libary? Let's explore!
```
Comment (sylvestre) [inline content/2023-06-02-shrinking-binaries.rst]:
```suggestion
          library? Let's explore!
```

File: content/2023-06-02-shrinking-binaries.rst [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+    $ LDFLAGS=-fuse-ld=lld\ -flto=thin CFLAGS=-O2\ -flto=thin CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                   102464
+
+Again, nothing surprising: having access to more information (through
+``-flto=full``) gives more optimization space than ``-flto=thin`` which
+(slightly) trades performance of the generated binary for memory usage of the
+actual compilation process. A trade-off we do not need to make for zlib but it's
+an other story for big project like Firefox.
```
Comment (sylvestre) [inline content/2023-06-02-shrinking-binaries.rst]:
```suggestion
another story for big project like Firefox.
```

File: content/2023-06-02-shrinking-binaries.rst [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+                     U __libc_start_main@GLIBC_2.34
+                     U perror@GLIBC_2.2.5
+                     U snprintf@GLIBC_2.2.5
+                     U strcmp@GLIBC_2.2.5
+                     U strlen@GLIBC_2.2.5
+                     U strrchr@GLIBC_2.2.5
+                     U unlink@GLIBC_2.2.5
+
+In addition to symbols from the libc, it uses a few symbols from zlib.
+This doesn't covert the whole zlib ABI though. Let's shrink the library just for our
```
Comment (sylvestre) [inline content/2023-06-02-shrinking-binaries.rst]:
```suggestion
This doesn't cover the whole zlib ABI though. Let's shrink the library just for our
```

File: content/2023-06-02-shrinking-binaries.rst [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+:date: 2023-06-02
+:category: mozilla
+:lang: en
+:authors: Serge Guelton
+:summary: What is the effect of different compiler flags on the size of a shared
+          libary? Let's explore!
+
+Recently, I've submitted a `patch
+<https://phabricator.services.mozilla.com/D179806>`_ that shaves ~2.5MB on one
+of the core Firefox library, ``libxul.so``. The patch is stupidly simple, but
```
Comment (sylvestre) [inline content/2023-06-02-shrinking-binaries.rst]:
```suggestion
of the core Firefox library, ``libxul.so``. The patch is remarkably simple, but
```

File: content/2023-06-02-shrinking-binaries.rst [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+    $ LDFLAGS=-fuse-ld=lld CFLAGS=-Oz CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                    81305
+
+.. code-block:: sh
+
+    $ LDFLAGS=-fuse-ld=lld CFLAGS=-O3 CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                   101943
```
Comment (lancesix) [inline content/2023-06-02-shrinking-binaries.rst]: I don't find this super easy to read as the information is spread across multiple lines.

Something like this makes it pop out a bit more:

```
echo "+----------+----------+" > result.txt
echo "| CFLAG    |     size |" >> result.txt
echo "+----------+----------+" >> result.txt

for opt in -O0 -O1 -O2 -O3 -Oz
do
        rm -rf _build
        mkdir -p _build
        (cd ./_build ; LDFLAGS=-fuse-ld=lld CFLAGS=$opt CC=clang ../configure && make && size -A libz.so)
        make -C _build -j32
        size -A _build/libz.so | awk -v cflag=$opt '/Total/ { printf "| %8s | %8d |\n", cflag, $2 }' >> result.txt
done
echo "+----------+----------+" >> result.txt
cat result.txt
```

which gives (with clang-14.0.0-1ubuntu1)  

```
+----------+----------+
| CFLAG    |     size |
+----------+----------+
|      -O0 |   141608 |
|      -O1 |    88164 |
|      -O2 |    99532 |
|      -O3 |   101644 |
|      -Oz |    81006 |
+----------+----------+
```

File: content/2023-06-02-shrinking-binaries.rst [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+Unsurprisingly ``-Oz`` yields the smaller binaries. ``-O3`` trades size for
+performance, and ``-O0`` performs a direct translation which uses a lot of
+space.
```
Comment (lancesix) [inline content/2023-06-02-shrinking-binaries.rst]: What about `-02` v.s. `-01`?

Also, I [believe ](https://clang.llvm.org/docs/CommandGuide/clang.html#cmdoption-o0) `-Oz` is `-O2` + extra options. Any chance that `-01` plus the same options can give smaller binaries (with some perf tradeoff)?

File: content/2023-06-02-shrinking-binaries.rst [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+    .data                       0   84304
+    .tm_clone_table             0   84304
+    .got.plt                  376   84304
+    .bss                        1   84680
+    .gnu.build.attributes     288       0
+    .comment                  110       0
+    Total                   72147
+
+We reduced the ``.eh_frame`` and ``.eh_frame_hdr`` sections at the expense of
+removing support for stack unwinding. Again, it's a trade-off but one we may want
```
Comment (lancesix) [inline content/2023-06-02-shrinking-binaries.rst:239]: and some debug-ability 😿

File: content/2023-06-02-shrinking-binaries.rst [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+                     U perror@GLIBC_2.2.5
+                     U snprintf@GLIBC_2.2.5
+                     U strcmp@GLIBC_2.2.5
+                     U strlen@GLIBC_2.2.5
+                     U strrchr@GLIBC_2.2.5
+                     U unlink@GLIBC_2.2.5
+
+In addition to symbols from the libc, it uses a few symbols from zlib.
+This doesn't covert the whole zlib ABI though. Let's shrink the library just for our
+usage using a version script that only references the symbol we want to use:
```
Comment (lancesix) [inline content/2023-06-02-shrinking-binaries.rst]:
```suggestion
usage using a version script that only references the symbols we want to use:
```

File: content/2023-06-02-shrinking-binaries.rst [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+    .dynamic                  432   59344
+    .got                       32   59776
+    .tm_clone_table             0   63904
+    .got.plt                  136   63904
+    .bss                        1   64040
+    .gnu.build.attributes     288       0
+    .comment                  110       0
+    Total                   51496
+
+The linker use the visibility information to iteratively remove code
```
Comment (lancesix) [inline content/2023-06-02-shrinking-binaries.rst]:
```suggestion
The linker uses the visibility information to iteratively remove code
```

File: content/2023-06-02-shrinking-binaries.rst [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+that's never referenced.
+
+
+Concluding Notes
+----------------
+
+Some of the remaining sections are purely informal. For instance:
```
Comment (lancesix) [inline content/2023-06-02-shrinking-binaries.rst]: Not sure about "informal".  Informational maybe?

**Pull request merged**